client/my-sites/patterns/components/pattern-gallery/client.tsx

```diff
@@ -14,12 +15,68 @@ import './style.scss';
 
 const LOGGED_OUT_USERS_CAN_COPY_COUNT = 3;
 
+type MasonryGalleryProps = PropsWithChildren< {
+	className?: string;
+	enableMasonry: boolean;
+} >;
+
+// Simulates a Masonry layout by applying negative `margin-top` on every item that doesn't sit in
+// the first row
+function MasonryGallery( { children, className, enableMasonry }: MasonryGalleryProps ) {
+	const ref = useRef< HTMLDivElement >( null );
+
+	useLayoutEffect( () => {
+		if ( ! ref.current || ! enableMasonry ) {
+			return;
+		}
+
+		const element = ref.current;
+
+		function calculateLayout() {
+			const columnCount = getComputedStyle( element ).gridTemplateColumns.split( ' ' ).length;
+
+			if ( columnCount === 1 ) {
+				return;
+			}
+
+			const items = [ ...element.querySelectorAll< HTMLElement >( '.pattern-preview' ) ];
```

Comment:
An opportunity to avoid this could be to have a composable set of components:

* `MasonryGallery` - the gallery wrapper
* `MasonryGalleryItem` - each pattern preview

That way we'll have the corresponding refs already there through the children elements, and that can make the recalculation more straightforward.

Reply:
I'm not that concerned with reusability here.